File: algorithm/leetcode/day17012026/WordSearch.java
public class WordSearch {
  // find path = dfs + graph search
  public boolean exist(char[][] board, String word) {
    int row = board.length;
    int col = board[0].length;
    for (int i = 0; i < row; i++) {
      for (int j = 0; j < col; j++) {
        // Start from the first meeting
        if (board[i][j] == word.charAt(0)) {
          if (dfs(board, word, row, col, i, j, 0)) {
            return true;
          }
        }
      }
    }

    return false;
  }

  boolean dfs(char[][] board, String word, int row, int col, int currRow, int currCol, int idx) {
    if (idx == word.length()) {
      return true;
    }
    if (currRow < 0 || currRow >= row || currCol < 0 || currCol >= col
        || board[currRow][currCol] != word.charAt(idx)) {
      return false;
    }

    // mark the column at visited
    char temp = board[currRow][currCol];
    board[currRow][currCol] = '*';

    boolean isExist = dfs(board, word, row, col, currRow - 1, currCol, idx + 1) ||
        dfs(board, word, row, col, currRow + 1, currCol, idx + 1) ||
        dfs(board, word, row, col, currRow, currCol - 1, idx + 1) ||
        dfs(board, word, row, col, currRow, currCol + 1, idx + 1);

    board[currRow][currCol] = temp;

    return isExist;
  }

  public static void main(String[] args) {
    WordSearch wordSearch = new WordSearch();
    // char[][] board = {{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
    // String word = "ABCCED";
    // System.out.println(wordSearch.exist(board, word));
    char[][] board = {{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
    String word = "ABCB";
    System.out.println(wordSearch.exist(board, word));
  }
}

File: interview/leetcode/day03032026/WordSearch.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board or not word:
            return False
        visited = set()

        def dfs(i: int, j: int, idx: int) -> bool:
            if idx == len(word):
                return True
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or (i*len(board[0]) + j) in visited or board[i][j] != word[idx]:
                return False

            visited.add(i*len(board[0]) + j)

            found = dfs(i-1, j, idx+1) or dfs(i, j+1, idx +
                                              1) or dfs(i+1, j, idx+1) or dfs(i, j-1, idx+1)

            visited.remove(i*len(board[0]) + j)

            return found

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    found = dfs(i, j, 0)

                    if found:
                        return True
        return False


sol = Solution()

print(sol.exist([["A", "B", "C", "E"], [
      "S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE"))
